include/Signal.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+#include <memory>
+#include <vector>
+#include <functional>
+
+namespace Framework
+{
+
+	template<typename> class CSignal;
+	template<typename T, typename... Args>
+	class CSignal<T (Args...)>
+	{
+
+	public:
+		class CConnection;
+
+		typedef std::shared_ptr<CConnection> CConnectionPtr;
+		typedef std::weak_ptr<CConnection> CConnectionWeakPtr;
+		typedef std::vector<CConnectionPtr> CConnectionListPtr;
+
+		typedef CSignal<T (Args...)> CSignalProto;
+		typedef std::function<void()> CConectionFunction;
+		typedef std::function<T (Args...)> CSlotFunction;
+		typedef std::weak_ptr<const CSlotFunction> CSlotFunctionPtr;
+
+		CSignal() = default;
+
+		CConnectionPtr Connect(const CSlotFunction& func, bool oneShot = false)
```

Comment:
This should probably be annotated with `nodiscard`. I was writing a test for this and I got caught. If the caller doesn't save the connection, it's not going to work. 

I'm not quite sure it's the best way to go around this... I can see why the signal doesn't keep strong references (less chances of leaking memory), but usage is a bit error prone. Anyways, I think it's better this way for now.

Reply:
that's the best way I could think of, the other way is to have a weak reference to the calling object, but that requires more changes to how the object was created.

`[[nodiscard]]` is C++17, should I apply it like https://github.com/jpd002/Play--Framework/blob/master/include/maybe_unused.h

Reply:
Yes.